File: go/conformance_test.go
package amino_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	amino "github.com/raiderrobert/amino/go"
)

// The corpus in ../spec/conformance is the specification. Cases whose xfail
// reason starts with "go:" are expected failures for this host; xfail reasons
// for other hosts are ignored and the case is required to pass here.

const corpusDir = "../spec/conformance"

type parseFile struct {
	Schema    string      `json:"schema"`
	Operators string      `json:"operators"`
	Cases     []parseCase `json:"cases"`
}

type parseCase struct {
	Name       string `json:"name"`
	Expression string `json:"expression"`
	Expect     string `json:"expect"`
	Type       string `json:"type"`
	Error      string `json:"error"`
	XFail      string `json:"xfail"`
}

type evalFile struct {
	Schema  string           `json:"schema"`
	Records []map[string]any `json:"records"`
	Cases   []evalCase       `json:"cases"`
}

type evalCase struct {
	Expression string  `json:"expression"`
	Matches    []int64 `json:"matches"`
	XFail      string  `json:"xfail"`
}

func xfailForGo(reason string) bool { return strings.HasPrefix(reason, "go:") }

func loadJSON(t *testing.T, path string, into any) {
	t.Helper()
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	if err := dec.Decode(into); err != nil {
		t.Fatalf("decode %s: %v", path, err)
	}
}

// normalize converts json.Number to int64 or float64, recursively, so records
// look the way application code would build them.
func normalize(v any) any {
	switch x := v.(type) {
	case json.Number:
		if i, err := x.Int64(); err == nil {
			return i
		}
		f, _ := x.Float64()
		return f
	case map[string]any:
		out := make(map[string]any, len(x))
		for k, e := range x {
			out[k] = normalize(e)
		}
		return out
	case []any:
		out := make([]any, len(x))
		for i, e := range x {
			out[i] = normalize(e)
		}
		return out
	}
	return v
}

func expandMacro(expr string) string {
	if strings.HasPrefix(expr, "@deep:") {
		n := 0
		for _, ch := range expr[len("@deep:"):] {
			n = n*10 + int(ch-'0')
		}
		return strings.Repeat("(", n) + "score > 1" + strings.Repeat(")", n)
	}
	return expr
}

func TestConformanceParse(t *testing.T) {
	files, _ := filepath.Glob(filepath.Join(corpusDir, "parse", "*.json"))
	if len(files) == 0 {
		t.Fatal("no parse corpus files found")
	}
	for _, path := range files {
		var f parseFile
		loadJSON(t, path, &f)
		var opts []amino.Option
		if f.Operators != "" {
			opts = append(opts, amino.WithOperators(f.Operators))
		}
		engine, err := amino.LoadSchema(f.Schema, opts...)
		if err != nil {
			t.Fatalf("%s: schema: %v", path, err)
		}
		for _, c := range f.Cases {
			c := c
			t.Run(filepath.Base(path)+"/"+c.Name, func(t *testing.T) {
				expr, err := engine.Parse(expandMacro(c.Expression))
				var failure string
				switch c.Expect {
				case "accept":
					if err != nil {
						failure = "expected accept, got " + err.Error()
					} else if expr.Type != c.Type {
						failure = "expected type " + c.Type + ", got " + expr.Type
					}
				case "reject":
					var ae *amino.Error
					if err == nil {
						failure = "expected reject with " + c.Error + ", got accept"
					} else if !errors.As(err, &ae) {
						failure = "expected *amino.Error, got " + err.Error()
					} else if string(ae.Code) != c.Error {
						failure = "expected code " + c.Error + ", got " + string(ae.Code)
					}
				default:
					t.Fatalf("bad expect %q", c.Expect)
				}
				if xfailForGo(c.XFail) {
					if failure == "" {
						t.Fatalf("marked xfail for go but passed; remove the marker: %s", c.XFail)
					}
					t.Skipf("expected failure: %s", c.XFail)
				}
				if failure != "" {
					t.Fatal(failure)
				}
			})
		}
	}
}

func TestConformanceEval(t *testing.T) {
	files, _ := filepath.Glob(filepath.Join(corpusDir, "eval", "*.json"))
	if len(files) == 0 {
		t.Fatal("no eval corpus files found")
	}
	for _, path := range files {
		var f evalFile
		loadJSON(t, path, &f)
		engine, err := amino.LoadSchema(f.Schema)
		if err != nil {
			t.Fatalf("%s: schema: %v", path, err)
		}
		records := make([]map[string]any, len(f.Records))
		for i, r := range f.Records {
			records[i] = normalize(r).(map[string]any)
		}
		for _, c := range f.Cases {
			c := c
			t.Run(filepath.Base(path)+"/"+c.Expression, func(t *testing.T) {
				var failure string
				compiled, err := engine.Compile([]amino.Rule{{ID: "q", Expr: c.Expression}}, nil)
				if err != nil {
					failure = "compile: " + err.Error()
				} else {
					results, err := compiled.Eval(records)
					if err != nil {
						failure = "eval: " + err.Error()
					} else {
						var got []int64
						for _, r := range results {
							if len(r.Matched) == 1 {
								got = append(got, r.ID.(int64))
							}
						}
						sort.Slice(got, func(i, j int) bool { return got[i] < got[j] })
						want := c.Matches
						if want == nil {
							want = []int64{}
						}
						if got == nil {
							got = []int64{}
						}
						if len(got) != len(want) {
							failure = "matches differ"
						} else {
							for i := range got {
								if got[i] != want[i] {
									failure = "matches differ"
								}
							}
						}
						if failure != "" {
							failure += ": want " + join(want) + ", got " + join(got)
						}
					}
				}
				if xfailForGo(c.XFail) {
					if failure == "" {
						t.Fatalf("marked xfail for go but passed; remove the marker: %s", c.XFail)
					}
					t.Skipf("expected failure: %s", c.XFail)
				}
				if failure != "" {
					t.Fatal(failure)
				}
			})
		}
	}
}

func join(ids []int64) string {
	parts := make([]string, len(ids))
	for i, id := range ids {
		parts[i] = jsonString(id)
	}
	return "[" + strings.Join(parts, ", ") + "]"
}

func jsonString(v any) string {
	b, _ := json.Marshal(v)
	return string(b)
}

File: typescript/test/conformance.test.ts
/**
 * Runs ../spec/conformance against this implementation. Cases whose xfail
 * reason starts with "typescript:" are expected failures here; other hosts'
 * reasons are ignored and the case must pass.
 */
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";

import { isAminoError, loadSchema, type Options } from "../src/index.js";

const CORPUS = join(__dirname, "..", "..", "spec", "conformance");

interface ParseCase {
  name: string;
  expression: string;
  expect: "accept" | "reject";
  type?: string;
  error?: string;
  xfail?: string;
}
interface ParseFile {
  schema: string;
  operators?: "standard" | "minimal";
  cases: ParseCase[];
}
interface EvalCase {
  expression: string;
  matches: number[];
  xfail?: string;
}
interface EvalFile {
  schema: string;
  records: Record<string, unknown>[];
  cases: EvalCase[];
}

function files(kind: string): string[] {
  const dir = join(CORPUS, kind);
  return readdirSync(dir)
    .filter((f) => f.endsWith(".json"))
    .sort()
    .map((f) => join(dir, f));
}

function load<T>(path: string): T {
  return JSON.parse(readFileSync(path, "utf8")) as T;
}

function expandMacro(expr: string): string {
  if (expr.startsWith("@deep:")) {
    const n = Number(expr.slice("@deep:".length));
    return "(".repeat(n) + "score > 1" + ")".repeat(n);
  }
  return expr;
}

const ours = (reason: string | undefined): boolean => reason?.startsWith("typescript:") ?? false;

/** Runs a check; under an xfail for this host, a pass is an error and a failure is expected. */
function guarded(xfail: string | undefined, check: () => void): void {
  if (!ours(xfail)) {
    check();
    return;
  }
  let failed = false;
  try {
    check();
  } catch {
    failed = true;
  }
  if (!failed) throw new Error(`marked xfail for typescript but passed; remove the marker: ${xfail}`);
}

describe("conformance: parse", () => {
  const paths = files("parse");
  expect(paths.length).toBeGreaterThan(0);
  for (const path of paths) {
    const f = load<ParseFile>(path);
    const options: Options = f.operators ? { operators: f.operators } : {};
    const engine = loadSchema(f.schema, options);
    describe(path.split("/").pop() as string, () => {
      for (const c of f.cases) {
        it(c.name, () => {
          guarded(c.xfail, () => {
            const result = engine.validate(expandMacro(c.expression));
            if (c.expect === "accept") {
              if (!result.ok) throw new Error(`expected accept, got ${result.error.message}`);
              expect(result.type).toBe(c.type);
            } else {
              if (result.ok) throw new Error(`expected reject with ${c.error}, got accept`);
              expect(isAminoError(result.error)).toBe(true);
              expect(result.error.code).toBe(c.error);
            }
          });
        });
      }
    });
  }
});

describe("conformance: eval", () => {
  const paths = files("eval");
  expect(paths.length).toBeGreaterThan(0);
  for (const path of paths) {
    const f = load<EvalFile>(path);
    const engine = loadSchema(f.schema);
    describe(path.split("/").pop() as string, () => {
      for (const c of f.cases) {
        it(c.expression, () => {
          guarded(c.xfail, () => {
            const compiled = engine.compile([{ id: "q", expr: c.expression }]);
            const got = compiled
              .eval(f.records)
              .filter((r) => r.matched.length === 1)
              .map((r) => r.id as number)
              .sort((a, b) => a - b);
            expect(got).toEqual(c.matches);
          });
        });
      }
    });
  }
});
